File: packages/open-memory/packages/openmemory-py/src/openmemory/core/vector/valkey.py

from typing import List, Optional, Dict, Any
import json
import logging
import asyncio
import numpy as np
from ..vector_store import VectorStore, VectorRow

logger = logging.getLogger("vector_store.valkey")

class ValkeyVectorStore(VectorStore):
    def __init__(self, url: str, prefix: str = "om:vec:"):
        self.url = url
        self.prefix = prefix
        self.client = None

    async def _get_client(self):
        import redis.asyncio as redis
        if not self.client:
            self.client = redis.from_url(self.url)
        return self.client

    def _key(self, id: str) -> str:
        return f"{self.prefix}{id}"

    async def storeVector(self, id: str, sector: str, vector: List[float], dim: int, user_id: Optional[str] = None):
        client = await self._get_client()
        key = self._key(id)
        vec_bytes = np.array(vector, dtype=np.float32).tobytes()

        mapping = {
            "id": id,
            "sector": sector,
            "dim": dim,
            "v": vec_bytes,
            "user_id": user_id or ""
        }
        await client.hset(key, mapping=mapping)

    async def getVectorsById(self, id: str) -> List[VectorRow]:
        client = await self._get_client()
        key = self._key(id)
        data = await client.hgetall(key)
        if not data: return []
        def dec(x): return x.decode('utf-8') if isinstance(x, bytes) else str(x)

        vec_bytes = data.get(b'v') or data.get('v')
        vec = list(np.frombuffer(vec_bytes, dtype=np.float32))

        return [VectorRow(
            dec(data.get(b'id') or data.get('id')),
            dec(data.get(b'sector') or data.get('sector')),
            vec,
            int(dec(data.get(b'dim') or data.get('dim')))
        )]

    async def getVector(self, id: str, sector: str) -> Optional[VectorRow]:
        rows = await self.getVectorsById(id)
        for r in rows:
            if r.sector == sector:
                return r
        return None

    async def deleteVectors(self, id: str):
        client = await self._get_client()
        await client.delete(self._key(id))

    async def search(self, vector: List[float], sector: str, k: int, filter: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:

        client = await self._get_client()
        query_vec = np.array(vector, dtype=np.float32)
        q_norm = np.linalg.norm(query_vec)

        cursor = 0
        results = []

        while True:
            cursor, keys = await client.scan(cursor, match=f"{self.prefix}*", count=100)
            if keys:
                pipe = client.pipeline()
                for key in keys:
                    pipe.hgetall(key)
                items = await pipe.execute()

                for item in items:
                    if not item: continue
                    def dec(x): return x.decode('utf-8') if isinstance(x, bytes) else str(x)

                    i_sector = dec(item.get(b'sector') or item.get('sector'))
                    if i_sector != sector: continue

                    if filter and filter.get("user_id"):
                        i_uid = dec(item.get(b'user_id') or item.get('user_id'))
                        if i_uid != filter["user_id"]: continue

                    v_bytes = item.get(b'v') or item.get('v')
                    v = np.frombuffer(v_bytes, dtype=np.float32)

                    dot = np.dot(query_vec, v)
                    norm = np.linalg.norm(v)
                    sim = dot / (q_norm * norm) if (q_norm * norm) > 0 else 0

                    results.append({
                        "id": dec(item.get(b'id') or item.get('id')),
                        "similarity": float(sim)
                    })

            if cursor == 0: break

        results.sort(key=lambda x: x["similarity"], reverse=True)
        return results[:k]

File: packages/open-memory/packages/openmemory-js/src/core/vector/valkey.ts
import { VectorStore } from "../vector_store";
import Redis from "ioredis";
import { env } from "../cfg";
import { vectorToBuffer, bufferToVector } from "../../memory/embed";

export class ValkeyVectorStore implements VectorStore {
    private client: Redis;

    constructor() {
        this.client = new Redis({
            host: env.valkey_host || "localhost",
            port: env.valkey_port || 6379,
            password: env.valkey_password,
            connectionName: "openmemory_vector_store_client",
        });
    }

    private getKey(id: string, sector: string): string {
        return `vec:${sector}:${id}`;
    }

    async storeVector(
        id: string,
        sector: string,
        vector: number[],
        dim: number,
        user_id?: string,
        project_id?: string,
    ): Promise<void> {
        const key = this.getKey(id, sector);
        const buf = vectorToBuffer(vector);

        await this.client.hset(key, {
            v: buf,
            dim: dim,
            user_id: user_id || "anonymous",
            project_id: project_id || "null",
            id: id,
            sector: sector,
        });
    }

    async deleteVector(id: string, sector: string): Promise<void> {
        const key = this.getKey(id, sector);
        await this.client.del(key);
    }

    async deleteVectors(id: string): Promise<void> {
        let cursor = "0";
        do {
            const res = await this.client.scan(
                cursor,
                "MATCH",
                `vec:*:${id}`,
                "COUNT",
                100,
            );
            cursor = res[0];
            const keys = res[1];
            if (keys.length) await this.client.del(...keys);
        } while (cursor !== "0");
    }

    async searchSimilar(
        sector: string,
        queryVec: number[],
        topK: number,
        user_id?: string,
        project_id?: string,
    ): Promise<Array<{ id: string; score: number }>> {
        // Valkey/Redis doesn't support user_id filtering in FT.SEARCH easily
        // For now we'll need to post-filter or use a more complex query
        const indexName = `idx:${sector}`;
        const blob = vectorToBuffer(queryVec);

        try {
            // Use FT.SEARCH with vector similarity
            const res = (await this.client.call(
                "FT.SEARCH",
                indexName,
                `*=>[KNN ${topK * 2} @v $blob AS score]`, // fetch more to allow filtering
                "PARAMS",
                "2",
                "blob",
                blob,
                "DIALECT",
                "2",
            )) as any[];

            // Parse results and filter by user_id if provided
            const results: Array<{ id: string; score: number }> = [];
            for (let i = 1; i < res.length; i += 2) {
                const key = res[i] as string;
                const fields = res[i + 1] as any[];
                let id = "";
                let dist = 0;
                let vec_user_id = "";
                let vec_project_id = "";

                for (let j = 0; j < fields.length; j += 2) {
                    if (fields[j] === "id") id = fields[j + 1];
                    if (fields[j] === "score") dist = parseFloat(fields[j + 1]);
                    if (fields[j] === "user_id") vec_user_id = fields[j + 1];
                    if (fields[j] === "project_id")
                        vec_project_id = fields[j + 1];
                }
                if (!id) id = key.split(":").pop()!;

                // Filter by user_id and project_id if provided
                const userMatch = !user_id || vec_user_id === user_id;
                const projectMatch =
                    !project_id ||
                    vec_project_id === project_id ||
                    vec_project_id === "system_global" ||
                    vec_project_id === "null" ||
                    vec_project_id === "";

                if (userMatch && projectMatch) {
                    results.push({ id, score: 1 - dist });
                    if (results.length >= topK) break;
                }
            }

            return results;
        } catch (e) {
            console.warn(
                `[Valkey] FT.SEARCH failed for ${sector}, falling back to scan (slow):`,
                e,
            );

            // Fallback: scan all vectors and filter
            let cursor = "0";
            const allVecs: Array<{
                id: string;
                vector: number[];
                user_id: string;
                project_id: string;
            }> = [];
            do {
                const res = await this.client.scan(
                    cursor,
                    "MATCH",
                    `vec:${sector}:*`,
                    "COUNT",
                    100,
                );
                cursor = res[0];
                const keys = res[1];
                if (keys.length) {
                    const pipe = this.client.pipeline();
                    keys.forEach((k) =>
                        pipe.hmget(k, "v", "user_id", "project_id"),
                    );
                    const buffers = await pipe.exec();
                    buffers?.forEach((b, idx) => {
                        if (b && b[1]) {
                            const [buf, vec_user_id, vec_project_id] = b[1] as [
                                Buffer,
                                string,
                                string,
                            ];
                            const id = keys[idx].split(":").pop()!;

                            // Filter by user_id and project_id during scan
                            const userMatch =
                                !user_id || vec_user_id === user_id;
                            const projectMatch =
                                !project_id ||
                                vec_project_id === project_id ||
                                vec_project_id === "system_global" ||
                                vec_project_id === "null" ||
                                vec_project_id === "";

                            if (userMatch && projectMatch) {
                                allVecs.push({
                                    id,
                                    vector: bufferToVector(buf),
                                    user_id: vec_user_id,
                                    project_id: vec_project_id,
                                });
                            }
                        }
                    });
                }
            } while (cursor !== "0");

            const sims = allVecs.map((v) => ({
                id: v.id,
                score: this.cosineSimilarity(queryVec, v.vector),
            }));
            sims.sort((a, b) => b.score - a.score);
            return sims.slice(0, topK);
        }
    }

    private cosineSimilarity(a: number[], b: number[]) {
        if (a.length !== b.length) return 0;
        let dot = 0,
            na = 0,
            nb = 0;
        for (let i = 0; i < a.length; i++) {
            dot += a[i] * b[i];
            na += a[i] * a[i];
            nb += b[i] * b[i];
        }
        return na && nb ? dot / (Math.sqrt(na) * Math.sqrt(nb)) : 0;
    }

    async getVector(
        id: string,
        sector: string,
    ): Promise<{ vector: number[]; dim: number } | null> {
        const key = this.getKey(id, sector);
        const res = await this.client.hmget(key, "v", "dim");
        if (!res[0]) return null;
        return {
            vector: bufferToVector(res[0] as unknown as Buffer),
            dim: parseInt(res[1] as string),
        };
    }

    async getVectorsById(
        id: string,
    ): Promise<Array<{ sector: string; vector: number[]; dim: number }>> {
        const results: Array<{
            sector: string;
            vector: number[];
            dim: number;
        }> = [];
        let cursor = "0";
        do {
            const res = await this.client.scan(
                cursor,
                "MATCH",
                `vec:*:${id}`,
                "COUNT",
                100,
            );
            cursor = res[0];
            const keys = res[1];
            if (keys.length) {
                const pipe = this.client.pipeline();
                keys.forEach((k) => pipe.hmget(k, "v", "dim"));
                const res = await pipe.exec();
                res?.forEach((r, idx) => {
                    if (r && r[1]) {
                        const [v, dim] = r[1] as [Buffer, string];
                        const key = keys[idx];
                        const parts = key.split(":");
                        const sector = parts[1];
                        results.push({
                            sector,
                            vector: bufferToVector(v),
                            dim: parseInt(dim),
                        });
                    }
                });
            }
        } while (cursor !== "0");
        return results;
    }

    async getVectorsBySector(
        sector: string,
    ): Promise<Array<{ id: string; vector: number[]; dim: number }>> {
        const results: Array<{ id: string; vector: number[]; dim: number }> =
            [];
        let cursor = "0";
        do {
            const res = await this.client.scan(
                cursor,
                "MATCH",
                `vec:${sector}:*`,
                "COUNT",
                100,
            );
            cursor = res[0];
            const keys = res[1];
            if (keys.length) {
                const pipe = this.client.pipeline();
                keys.forEach((k) => pipe.hmget(k, "v", "dim"));
                const res = await pipe.exec();
                res?.forEach((r, idx) => {
                    if (r && r[1]) {
                        const [v, dim] = r[1] as [Buffer, string];
                        const key = keys[idx];
                        const id = key.split(":").pop()!;
                        results.push({
                            id,
                            vector: bufferToVector(v),
                            dim: parseInt(dim),
                        });
                    }
                });
            }
        } while (cursor !== "0");
        return results;
    }
}
